### ISSUE TITLE
Alluxio S3 Download via FUSE 6x as slow compared to first loading in Alluxio and then downloading via FUSE

### ISSUE THREAD

**Issue by peterroelants** [Alluxio/alluxio#13161]
**Alluxio Version:**
2.5.0-1

**Describe the bug**
Downloading a file from S3 that is not in Alluxio (+3 min) is 6 times as slow as first loading the same file in Alluxio (11 s) and then downloading it via FUSE (27 s).

**To Reproduce**
I have a 770.7 MB file in Amazon S3 of which the bucket is mounted in Alluxio. I tested the duration of the following running Alluxio on EC2 in the same geographical zone as the S3 bucket:

- Download directly from S3 with AWS CLI: 11 seconds
    - `time aws s3 cp s3://<bucket_name>/<file_name> .`
- Load file into Alluxio from worker node: 11 seconds
  - `time alluxio fs load /<alluxio_dir>/<file_name>`
- Download file via Alluxio FUSE (file is Already in Worker cache via `load`): 27 seconds
  - `time cat ${MOUNT_POINT}/<alluxio_dir>/<file_name> > /dev/null`
- Download file via Alluxio FUSE (file is not yet in Worker): **> 3 minutes** ❗
  - `time cat ${MOUNT_POINT}/<alluxio_dir>/<file_name> > /dev/null`

**Expected behavior**
Downloading via Alluxio FUSE taking approximately as long as first downloading to Alluxio Worker and then downloading via FUSE.

**Urgency**
High:
This bug makes S3 Access from Alluxio FUSE extremely slow. If I run the same on my local machine, not on EC2, it takes more than an hour to download the same file via Alluxio FUSE, while it only takes 70s with S3 CLI.

**Additional context**
- Alluxio run on in Docker on `t3.xlarge` EC2 instance
- FUSE config:
```
alluxio.fuse.jnifuse.enabled=true
alluxio.user.file.readtype.default=CACHE
alluxio.user.file.writetype.default=ASYNC_THROUGH
alluxio.fuse.maxwrite.bytes=131072
alluxio.fuse.shared.caching.reader.enabled=false
alluxio.user.metadata.cache.enabled=false
```

**Comment by peterroelants** [comment]
To give some non EC2 numbers when I run the same on my local machine:

- Download directly from S3 with boto3: 70 sec
- Load file into Alluxio from worker node: 2 min 30 sec
- Download file via Alluxio FUSE (file is Already in Worker cache via load): 14 sec
- Download file via Alluxio FUSE (file is not yet in Worker): **> 1 hour (hangs )❗** 

Notice that now downloading from S3 via Alluxio Worker becomes twice as slow compared to downloading with the boto3 AWS API. And that downloading via FUSE becomes extremely slow when the file is not yet cached.

Besides slowness I'm also seeing a lot of following warnings in the logs on my local machine:
- worker logs:
   - [Received DATA frame for an unknown stream](https://github.com/Alluxio/alluxio/issues/13138)
   - ` WARN  BlockReadHandler - Getting buffer for remote read took longer than 10000 ms. reader: alluxio.worker.block.UnderFileSystemBlockReader location: s3://<bucket>/<file> bytes: 1048576 openMs: 0 transferMs: 10486 durationMs: 10486`
 - FUSE logs:
   - `WARN  AlluxioJniFuseFileSystem - read(path=<alluxio_path>,buf=java.nio.DirectByteBuffer[pos=131072 lim=131072 cap=131072],size=131072,offset=549437440) returned 131072 in 11350 ms (>=10000 ms)`

**Comment by peterroelants** [comment]
The issue seems to be largely related to downloading large files from Alluxio.

If I store the same file as small files (chunked format), the FUSE read only takes *2.5 seconds* (instead of > 3 minutes).

**Comment by peterroelants** [comment]
If I store the same file as small files (chunked format), the FUSE read takes **1 minute** (instead of > 3 minutes).

**Comment by peterroelants** [comment]
I did some more tests related to this, and got some interesting results when timing data access. I run these tests on Amazon EC2 with files stored on S3 in 2 versions:
- One single large file with 770MB of data: `<file_name>`
- The same 770MB data split into different chunks spread over different smaller files: `<file_chunked>/<chunks>`

While testing direct access with [`cat`](https://man7.org/linux/man-pages/man1/cat.1.html), [Python boto3](https://boto3.amazonaws.com/v1/documentation/api/latest/index.html). And indirect access with [Python Dask](https://dask.org/) reading an Alluxio FUSE mounted filesystem, and with Dask via [fsspec](https://github.com/intake/filesystem_spec).

- Read from FUSE when NOT in cache with `cat ${MOUNT_POINT}/<alluxio_dir>/<file_name> > /dev/null`:
  - **3m30.949s**
- Read from FUSE when in cache with `cat ${MOUNT_POINT}/<alluxio_dir>/<file_name> > /dev/null`: 
  - **23.701s**
- Python boto3 direct download (no Alluxio) of `/<alluxio_dir>/<file_name>`:
  - **4.989s**
- Dask Alluxio FUSE read when NOT in cache of `/<alluxio_dir>/<file_name>` (will have Dask overhead):
  - **27.152s**
- Dask Alluxio FUSE read when in cache of `/<alluxio_dir>/<file_name>` (will have Dask overhead):
  - **2.992s**
- Dask Alluxio FUSE read when NOT in cache of `/<alluxio_dir>/<file_chunked>/<chunks>` (will have Dask overhead):
  - **10.422s**
- Dask Alluxio FUSE read when in cache of `/<alluxio_dir>/<file_chunked>/<chunks>` (will have Dask overhead):
  - **2.974s**
- Dask fsspec S3 read (no Alluxio) of `/<alluxio_dir>/<file_chunked>/<chunks>`(will have Dask overhead):
  - **7.199s**
  
  What is interesting:
  - Alluxio FUSE does not play well with `cat` if the files are not in Alluxio for some reason (the difference is so big I'm suspecting this is a _bug_?)
  - Reading small S3 files via Alluxio FUSE is 3x as fast as reading the same data as a single file in S3.

**Comment by peterroelants** [comment]
I did some more tests on JNR vs JNI FUSE when reading files from Alluxio FUSE (**all in worker cache**).

| Description | Dask single | `cat` single | Dask chunked | `cat` chunked |
|:----------- |:-----------:|:------------:|:------------:|:-------------:|
| 1 - JNR-FUSE, apt fuse, no optimizations | 627 ms | 433 ms | 840 ms | 480 ms |
| 2 - JNR-FUSE, patched fuse, optimized | 621 ms | 436 ms | 748 ms | 514 ms |
| 3 - JNI-FUSE, apt fuse, no optimizations | 714 ms | 14.6 s | 875 ms | 12.8 s |
| 4 - JNI-FUSE, patched fuse, no optimizations | 716 ms | 12.9 s | 943 ms | 14.9 s |
| 4 - JNI-FUSE, patched fuse, optimized | **243 ms** | **100 ms** | **394 ms** | **138 ms** |

Where:
- Either files are read via Python [Dask](https://dask.org/) or [`cat`](https://man7.org/linux/man-pages/man1/cat.1.html).
- Both a single large file ("single") or the same data split into different chunks spread over different smaller files ("chunked") are tested.
- "apt fuse" is fuselib installed via apt (fusermount version: 2.9.9)
- "patched fuse" is the [patched libfuse](https://github.com/cheyang/libfuse/tree/fuse_2_9_5_multi_threads) used in the Alluxio Docker container
- "no optimizations" is fuse mounted with the default settings
- "optimized" is fuse mounted with the [recommended optimization settings](https://docs.alluxio.io/os/user/stable/en/api/POSIX-API.html#configure-mount-point-options#:~:text=Tuning%20mount%20options).

Some things to notice:
- Reading from JNI-FUSE with `cat` without optimized settings is really slow (order of seconds iso milliseconds), to the point where I suspect a bug!
- The [recommended optimization settings](https://docs.alluxio.io/os/user/stable/en/api/POSIX-API.html#configure-mount-point-options#:~:text=Tuning%20mount%20options) result in a 3-4x speedup compared to the default settings.

**Comment by peterroelants** [comment]
I think this issue is multiple issue in one. I extracted out the slow reading with `cat` into a separate issue: https://github.com/Alluxio/alluxio/issues/13206
